Diurnal cycle climatology

chengzhuzhang opened the pull request:
This PR is to add diurnal cycle phase map:
* TRMM 0.25 degree data is pre-processed for monthly diurnal cycle and added.
*Default regions include: "CONUS", "20S20N", "W_Pacific", "Amazon","global","50S50N"
*example results: https://portal.nersc.gov/cfs/e3sm/zhang40/tests/diurnal_cycle_climo/viewer/diurnal_cycle/index.html
**updated results** with `normalize_test_amp = False`: https://portal.nersc.gov/cfs/e3sm/zhang40/tests/diurnal_cycle_climo_no_scale_test/viewer/

To post-process model output files to create monthly diurnal cycle that E3SM Diags could read, use the new NCO sub-monthly capability. Example scripts on Cori:
```
#!/bin/bash
  
source /global/cfs/cdirs/e3sm/software/anaconda_envs/load_latest_e3sm_unified.sh

# Low-res Cori simulations
drc_in=/global/cfs/cdirs/e3smpub/E3SM_simulations/20180215.DECKv1b_H1.ne30_oEC.edison/archive/atm/hist

drc_out=/global/cfs/cdirs/e3sm/zhang40/nco_scripts/results/diurnal_climo/native
drc_rgr=/global/cfs/cdirs/e3sm/zhang40/nco_scripts/results/diurnal_climo/rgr
caseid=20180215.DECKv1b_H1.ne30_oEC.edison.cam.h4

#cd ${drc_in};ls ${caseid}*.nc | ncclimo --var=PRECT --clm_md=hfc --caseid=$caseid --ypf=1 --yr_srt=2000 --yr_end=2004 --drc_out=${drc_out} -O $drc_rgr --map=/global/homes/z/zender/data/maps/map_ne30np4_to_cmip6_180x360_aave.20181001.nc
cd ${drc_in};ls ${caseid}*200*.nc | ncclimo --clm_md=hfc --caseid=$caseid --ypf=1 --yr_srt=2000 --yr_end=2009 --drc_out=${drc_out} -O $drc_rgr --map=/global/homes/z/zender/data/maps/map_ne30np4_to_cmip6_180x360_aave.20181001.nc
```

Notes here for improvements:
Use same scale for test and ref data, and to make a parameter so that user can choose to use different scales for test and ref. 
The CONUS domain range needs to be revised.

chengzhuzhang commented:
A new parameter, `normalize_test_amp` (default setting is `True`) has been added so that an option to plot test and ref on different scales are allowed. The CONUS range is extended. 

Note, this PR is to support climo diurnal cycle files, time series support will be in a future PR. Test will be added once we figure out a way to host test data.

forsyth2 commented on acme_diags/driver/utils/diurnal_cycle.py:
```diff
@@ -0,0 +1,152 @@
+
+        var_season = ma.zeros([ncycle]+[len(idx[0])]+list(numpy.shape(v))[1:])
+        var_season[n,] = v[idx]
+    var_daily = numpy.reshape(var_season,(ncycle,int(var_season.shape[1]/time_freq),time_freq,var_season.shape[2],var_season.shape[3]))
```
Did you check that the `reshape` returns the proper results? For instance, in Streamflow Diags, there was a call to `reshape` creating rows of `[January of year 1, February of year 1,...]` rather than `[January of year 1, January of year 2,...]`.

forsyth2 commented on acme_diags/driver/utils/diurnal_cycle.py:
```diff
@@ -0,0 +1,152 @@
+        var_season = ma.zeros([ncycle]+[len(idx[0])]+list(numpy.shape(v))[1:])
+        var_season[n,] = v[idx]
+    var_daily = numpy.reshape(var_season,(ncycle,int(var_season.shape[1]/time_freq),time_freq,var_season.shape[2],var_season.shape[3]))
+    var_diurnal = ma.average(var_daily,axis=1).squeeze()
```
Is `axis=1` the correct axis to average over?

forsyth2 reviewed (commented):
- Mostly just cleanup comments.
- I tried following the logic of `acme_diags/driver/diurnal_cycle_driver.py` and left a couple comments. I think it would be beneficial to add some comments in the code explaining what's going on. There's a lot of array shape/index manipulation.
- I would also squash the 11 commits down to 1. 
- I didn't run tests since you mentioned those would be done later.

chengzhuzhang commented:
I should have included the updated figures: updated results with normalize_test_amp = False: https://portal.nersc.gov/cfs/e3sm/zhang40/tests/diurnal_cycle_climo_no_scale_test/viewer/

chengzhuzhang commented on acme_diags/driver/utils/diurnal_cycle.py:
```diff
@@ -0,0 +1,152 @@
+import MV2
+
+
+def composite_diurnal_cycle(var, season):
```
I intend to keep the driver clean and and utilities in one place.

chengzhuzhang commented:
> I should have included the updated figures: updated results with `normalize_test_amp = False`: https://portal.nersc.gov/cfs/e3sm/zhang40/tests/diurnal_cycle_climo_no_scale_test/viewer/
> 
> The default output that normalizes test data amplitude: https://portal.nersc.gov/cfs/e3sm/zhang40/tests/diurnal_cycle_climo_scale_test/viewer/

@golaz Hi Chris, I'm about merging the diurnal cycle PR which supports climatology diurnal cycle files. It would be helpful if you can visual check the results from above link to see if any changes are desired.

@forsyth2 I should have addressed most of the review comments and added comments on code blocks to add clarity. Let me know if they look okay.

forsyth2 commented:
@chengzhuzhang Thanks for the updated plots. Those look better.

Did you push your latest changes to GitHub? The last commit I see is from before I reviewed.

chengzhuzhang commented:
@forsyth2 It's now pushed in...

forsyth2 commented on acme_diags/driver/utils/diurnal_cycle.py:
```diff
@@ -56,11 +56,13 @@ def composite_diurnal_cycle(var, season):
+        # Get time index for each month/season. 
+        idx = numpy.array([season_idx[cycle[n]][var_time_absolute[i].month-1]
+                          for i in range(len(var_time_absolute))], dtype=numpy.int).nonzero()
+        # var_season has shape (ncycle, ntimesteps, [lat,lon]) 
```
I'm a bit confused as to how `len(idx[0])` is `ntimesteps`. Say the first season is `'DJF'`. If `n == 0`, then `cycle[n] == cycle[0] == 'DJF'`. Then, `season_idx[cycle[n]] == season_index['DJF'] == [1,1,0,0,0,0,0,0,0,0,0,1]`. Initially, the list comprehension has `i == 0`, so if `var_time_absolute[0].month == '2'` for  February, then we would take `2-1=1` as the index. That gives us `idx[0] == season_idx['DJF'][1] == 1`. It seems to me like `ntimesteps` is actually equal to `len(idx)` then. Am I not following correctly? Thanks.

chengzhuzhang replied:
Yes it is a bit confusing. `nonzero()` return a tuple of array, so len(idx[0]) gives the actual `ntimesteps`. i.e., len(idx)= 1 (only one element in that tuple), len(idx[0]) equals to 8, which is the number of time steps, given climatology diurnal cycle of one season.

forsyth2 replied:
Oh, that makes sense. Thank you!

chengzhuzhang replied:
Thank you for the careful review.

forsyth2 reviewed (commented):
Thanks for updating. I'm still a bit confused on some of the array shapes, but if the generated plots look correct, then I suppose everything is in working order.
